logproof/src/bfv_statement.rs

```diff
@@ -0,0 +1,861 @@
+///                                               [   u_2  ]
+///                                               [ e_1[0] ]
+///                                               [ e_2[0] ]
+///                                               [ e_1[1] ]
+///                                               [ e_2[1] ]
+///                                               [   -s   ]
+///                                               [  -e_3  ]
+///
+/// ```
+///
+/// # Remarks
+/// The remainder for a given (plaintext, plaintext modulus, ciphertext modulus) trio should be
+/// constant, and thus it should technically be possible to reuse the remainder for multiple
+/// encryptions (public or private) of a single plaintext message, like we do for the delta scaling
+/// parameter. However, since the remainder is held in each [`BfvWitness`], I've gone with the less
+/// surprising implementation where we have a remainder witness for each statement.
+pub fn generate_prover_knowledge<C, P, S, T, B, const N: usize>(
+    statements: &[BfvProofStatement<C, P>],
+    messages: &[Plaintext], // may want messages AsRef as well.. we'll see
+    witness: &[BfvWitness<S>],
+    params: &T,
+    ctx: &Context,
+) -> LogProofProverKnowledge<Z<N, B>>
+where
+    B: BarrettConfig<N>,
+    C: AsRef<Ciphertext>,
+    P: AsRef<PublicKey>,
+    S: AsRef<SecretKey>,
+    T: StatementParams,
+{
+    let vk = generate_verifier_knowledge(statements, params, ctx);
+
+    let s = compute_s(statements, messages, witness);
+
+    LogProofProverKnowledge { vk, s }
+}
+
+/// Generate only the [`LogProofVerifierKnowledge`] for a given set of [`BfvProofStatement`]s.
+///
+/// See the documentation for [`generate_prover_knowledge`] for more information.
+pub fn generate_verifier_knowledge<C, P, T, B, const N: usize>(
+    statements: &[BfvProofStatement<C, P>],
+    params: &T,
+    ctx: &Context,
+) -> LogProofVerifierKnowledge<Z<N, B>>
+where
+    B: BarrettConfig<N>,
+    C: AsRef<Ciphertext>,
+    P: AsRef<PublicKey>,
+    T: StatementParams,
+{
+    let a = compute_a(statements, params, ctx);
+    let t = compute_t(statements, ctx);
+    let bounds = compute_bounds(statements, params);
+    let f = Polynomial {
+        coeffs: {
+            let degree = params.degree() as usize;
+            let mut cs = vec![Zq::zero(); degree + 1];
+            cs[0] = Zq::one();
+            cs[degree] = Zq::one();
+            cs
+        },
+    };
+
+    LogProofVerifierKnowledge { a, t, bounds, f }
+}
+
+fn compute_a<C, P, T, B, const N: usize>(
+    statements: &[BfvProofStatement<C, P>],
+    params: &T,
+    ctx: &Context,
+) -> PolynomialMatrix<Z<N, B>>
+where
+    B: BarrettConfig<N>,
+    C: AsRef<Ciphertext>,
+    P: AsRef<PublicKey>,
+    T: StatementParams,
+{
+    let mut offsets = IdxOffsets::new(statements);
+    let (rows, cols) = offsets.a_shape();
+    let mut a = PolynomialMatrix::new(rows, cols);
+
+    let d = Polynomial {
```

Comment:
Is this proving textbook BFV, or SEAL's variant that Ryan got working?

Reply:
Ryan's - much of the `seal_fhe -> sunscreen_math` conversions are copied from e.g. `logproof::test`. (Which can be removed after we adapt `sunscreen_runtime` to use `logproof::bfv_statement`).